bluetooth: host: Report status of CS procedure complete events

ryanjh opened the pull request:
Adds status into notify parameter:
- LE CS Read Remote Supported Capabilities Complete event
- LE CS Read Remote FAE Table Complete event
- LE CS Config Complete event
- LE CS Procedure Enable Complete event

ryanjh committed:
bluetooth: host: Report status of Channel Sounding complete events

If the HCI status of a complete event is not BT_HCI_ERR_SUCCESS,
the remaining parameters could be invalid.

In this case, the params is passed as NULL pointer to the callbacks.
- LE CS Read Remote Supported Capabilities Complete event
- LE CS Read Remote FAE Table Complete event
- LE CS Config Complete event
- LE CS Security Enable Complete event
- LE CS Procedure Enable Complete event

This change avoids forwarding the invalid fileds to the applications.

Signed-off-by: Ryan Chu <ryan.chu@nordicsemi.no>

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -814,12 +813,14 @@ void bt_hci_le_cs_config_complete_event(struct net_buf *buf)
 
-	if (evt->action == BT_HCI_LE_CS_CONFIG_ACTION_REMOVED) {
+	if (evt->action == BT_HCI_LE_CS_CONFIG_ACTION_REMOVED &&
+	    evt->status == BT_HCI_ERR_SUCCESS) {
```
it can't happen that action = 0 and status != 0 ?

olivier-le-sage commented:
As the spec states

> Vol 4, Part E, 4.5 Command error handling
> If an error occurs for a command for which an HCI_Command_Complete event
> is returned, the Return Parameters field may only contain some of the return
> parameters specified for the command. [...]  If there is a Handle
> parameter or a BD_ADDR parameter right after the Status parameter, this parameter
> shall also be returned so that the Host can identify to which instance of a command
> the HCI_Command_Complete event belongs. In this case, the Handle or BD_ADDR
> parameter shall have exactly the same value as that in the corresponding command
> parameter. It is implementation specific whether more parameters will be returned in
> case of an error; if they are not, the event will be shorter than if they were.
> ...
> The above also applies to commands that have associated command specific
> completion events with a Status parameter in their completion event, with the
> exceptions shown in Table 4.1,
> Event Valid parameters

and none of these control procedures here are present in Table 4.1, I believe we shouldn't read the invalid fields.

jhedberg commented:
I'd consider passing `NULL` and a separate `uint8_t status` parameter to the callbacks, to avoid any chance of access to the other struct fields in the case of errors.

ryanjh commented:
> I'd consider passing `NULL` and a separate `uint8_t status` parameter to the callbacks, to avoid any chance of access to the other struct fields in the case of errors.

Is my understanding correct that the suggestion including passing status and NULL pointer to the notify APIs?

jhedberg commented:
> Is my understanding correct that the suggestion including passing status and NULL pointer to the notify APIs?

Yes, that's what I meant. However, do we really need this absolutely everywhere - even where there are not parameters. I'm thinking of something like `security_enabled()`. Feels a bit unintitive to have a call named like that, yet it might mean that security wasn't enabled. If you really do want to indicate a security enabling failure, perhaps a separate `security_failed()` would make sense, or maybe rename the existing callback to `security_update()` (just brainstorming here).

ryanjh commented:
How about renaming as below, then the application handles CS callbacks in a similar way? 
le_cs_remote_capabilities_available -> le_cs_read_remote_capabilities_complete
le_cs_remote_fae_table_available     -> le_cs_read_remote_fae_table_complete
le_cs_security_enabled                      -> le_cs_security_enable_complete
le_cs_procedure_enabled                  -> le_cs_procedure_enable_complete

By taking this chance, could we consider removing le_cs_config_removed?
le_cs_config_created -> le_cs_config_complete

jhedberg commented:
@ryanjh good suggestions, though I'm not quite sure of the intuitiveness or ease of use (from an app perspective) of the config_complete thing. Note: The Zephyr Bluetooth API is *not* an HCI API or even an HCI wrapper. In fact, Zephyr has even supported non-HCI implementations of the API in the past (i.e. something else than the Zephyr host stack). To some extent, the CS APIs feels a bit too much like HCI to me, and it would be good to look for ways to provide a bit more app-friendly and higher level abstractions to what's happening underneath.

ryanjh commented:
Thanks for the feedback. Another option could be adding more callbacks for the case of errors. For instance, le_cs_read_remote_capabilities_failed, le_cs_read_remote_fae_table_failed, ... le_cs_procedure_enable_failed.
It might not be ideal to introduce more CS callbacks before abstraction API is in place.

To align callbacks with current CS API implementation, I will update this PR with renaming as the first attempt.

ryanjh committed:
bluetooth: host: Apply callback changes of CS complete events

The CS complete callbacks provide both status and params.
In the case of errors, NULL pointer is passed to the params of callbacks.

Signed-off-by: Ryan Chu <ryan.chu@nordicsemi.no>
